### PR TITLE
refactor: give Pipeline a database

### PR DESCRIPTION
**Pull request by mattsse**
blocked by #2550

this refactors the Pipeline and how it's invoked

* give Pipeline ownership of Database, this simplifies how the pipeline is called
* update all breaking changes database types

There are likely some additional simplifications we can do now but will address them when I see them

### CODE REVIEW

File: crates/net/downloaders/src/bodies/bodies.rs [hunk 1]
```diff
@@ -249,7 +249,7 @@ where
 impl<B, DB> BodiesDownloader<B, DB>
 where
     B: BodiesClient + 'static,
-    DB: Database,
+    DB: Database + Unpin + 'static,
```
Comment (gakonst) [inline crates/net/downloaders/src/bodies/bodies.rs:252]: is this static bound OK for library use?

**Review by gakonst — APPROVED** [review]
fine w/ me, pretty simple

File: crates/net/downloaders/src/bodies/bodies.rs (hunk 1 above)
Comment (mattsse) [reply crates/net/downloaders/src/bodies/bodies.rs:252]: yeh, it's just a type bound

**Review by rkrasiuk — APPROVED** [review]
love it!

File: crates/stages/src/pipeline/mod.rs [hunk 2]
```diff
@@ -169,28 +147,28 @@ where
     /// Consume the pipeline and run it until it reaches the provided tip, if set. Return the
     /// pipeline and its result as a future.
     #[track_caller]
-    pub fn run_as_fut(mut self, db: Arc<DB>, tip: Option<H256>) -> PipelineFut<DB> {
+    pub fn run_as_fut(mut self, tip: Option<H256>) -> PipelineFut<DB> {
         // TODO: fix this in a follow up PR. ideally, consensus engine would be responsible for
         // updating metrics.
-        self.register_metrics(db.clone());
+        self.register_metrics();
```
[the remark this answers is not in the document]
Comment (gakonst) [reply crates/stages/src/pipeline/mod.rs:153]: we can do in follow-up, merging

**Pull request merged**